lib/pos_network.go

```diff
@@ -0,0 +1,406 @@
+	// sign the (ProposedInView, BlockHash) pair for the block.
+	//
+	// From the block hash, we can look up the block height, the validator set at that
+	// block height, and the ordering of validators in consensus which identifies the
+	// present signers in the provided signers list. We can then use this to determine
+	// if the QC has 2/3rds of the total stake.
+	ValidatorsVoteAggregatedSignature *AggregatedBLSSignature
+}
+
+// Performs a deep equality check between two QuorumCertificates, and returns
+// true if the values of the two are identical.
+func (qc *QuorumCertificate) Eq(other *QuorumCertificate) bool {
+	if qc == nil && other == nil {
+		return true
+	}
+
+	if (qc == nil) != (other == nil) {
+		return false
+	}
+
+	return bytes.Equal(qc.BlockHash.ToBytes(), other.BlockHash.ToBytes()) &&
+		qc.ProposedInView == other.ProposedInView &&
+		qc.ValidatorsVoteAggregatedSignature.Eq(other.ValidatorsVoteAggregatedSignature)
+}
+
+func (qc *QuorumCertificate) ToBytes() ([]byte, error) {
+	retBytes := []byte{}
+
+	// BlockHash
+	if qc.BlockHash == nil {
+		return nil, errors.New("QuorumCertificate.ToBytes: BlockHash must not be nil")
+	}
+	retBytes = append(retBytes, qc.BlockHash.ToBytes()...)
+
+	// ProposedInView
+	retBytes = append(retBytes, UintToBuf(qc.ProposedInView)...)
+
+	// ValidatorsVoteAggregatedSignature
+	if qc.ValidatorsVoteAggregatedSignature == nil {
+		return nil, errors.New("QuorumCertificate.ToBytes: ValidatorsVoteAggregatedSignature must not be nil")
+	}
+	encodedValidatorsVoteAggregatedSignature, err := qc.ValidatorsVoteAggregatedSignature.ToBytes()
+	if err != nil {
+		return nil, errors.Wrapf(err, "QuorumCertificate.ToBytes: Error encoding ValidatorsVoteAggregatedSignature")
+	}
+	retBytes = append(retBytes, encodedValidatorsVoteAggregatedSignature...)
+
+	return retBytes, nil
+}
+
+func DecodeQuorumCertificate(rr io.Reader) (*QuorumCertificate, error) {
+	var qc QuorumCertificate
+	var err error
+
+	qc.BlockHash, err = ReadBlockHash(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DecodeQuorumCertificate: Error decoding BlockHash")
+	}
+
+	qc.ProposedInView, err = ReadUvarint(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DecodeQuorumCertificate: Error decoding ProposedInView")
+	}
+
+	qc.ValidatorsVoteAggregatedSignature, err = DecodeAggregatedBLSSignature(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DecodeQuorumCertificate: Error decoding ValidatorsVoteAggregatedSignature")
+	}
+
+	return &qc, nil
+}
+
+// This is an aggregated BLS signature from a set of validators. Each validator's
+// presence in the signature is denoted in the provided signers list. I.e. if the
+// list's value at index 0 is 1, then the validator identified by that index is
+// present in the aggregated signature. The indices of all validators are expected
+// to be known by the caller.
+type AggregatedBLSSignature struct {
+	// TODO: Switch this to a bitlist, which will result in ~8x reduction in total
```

Comment:
you'll tackle this in a new PR, right?

Reply:
Yes, it's addressed in PR https://github.com/deso-protocol/core/pull/565 further down the PR stack.